### PR TITLE
HHH-13496 Performance focused improvements from some static code analysis

### PR DESCRIPTION
**Pull request by Sanne**
hi @sebersole , @dreab8 , this PR is rather large but there's a lot of small performance improvements.

I'd push it already as I'm quite confident about it, but I don't wish it to get in the way of any big refactoring.

Could you just skim through the list of files it touches? I'm happy to apply just a subset of commits if that's more convenient for you.

Thanks

 - https://hibernate.atlassian.net/browse/HHH-13496

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/cfg/CopyIdentifierComponentSecondPass.java [hunk 1]
```diff
@@ -79,8 +79,8 @@ public void doSecondPass(Map persistentClasses) throws MappingException {
 		AtomicInteger index = new AtomicInteger( 0 );
 		if ( columnByReferencedName.isEmpty() ) {
 			isExplicitReference = false;
-			for (Ejb3JoinColumn joinColumn : joinColumns) {
-				columnByReferencedName.put( "" + index.get(), joinColumn );
+			for ( Ejb3JoinColumn joinColumn : joinColumns ) {
+				columnByReferencedName.put( "0", joinColumn );
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/cfg/CopyIdentifierComponentSecondPass.java]: the index is incremented inside the for loop, if `joinColumns`contains more than 1 `Ejb3JoinColumn` with this change we are not using the correct value for the Map key.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/cfg/CopyIdentifierComponentSecondPass.java]: I think the change should be `columnByReferencedName.put(  String.valueOf( index.get() ) , joinColumn );`

**Comment by dreab8** [comment]
@Sanne apart for the comment it looks fine to me, not easy to choose only subset of the commits so if for @sebersole it is fine I would apply all the commits.

File: hibernate-core/src/main/java/org/hibernate/cfg/CopyIdentifierComponentSecondPass.java (hunk 1 above)
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/cfg/CopyIdentifierComponentSecondPass.java]: nice catch. Not sure how I missed the loop.

Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/cfg/CopyIdentifierComponentSecondPass.java]: fixed

**Comment by Sanne** [comment]
@sebersole I've been hesitant to push this to not "get in the way" of your larger refactoring, but thinking about it you could just use a "ours/theirs" strategy when rebasing to discard any changes I made which conflicts with any of yours.

So it's easier if I just push this optimistically; if this ever gets in your way let me know I can help rebasing.

**Comment by Sanne** [comment]
merged

**Pull request closed without merging**